### PR TITLE
core-to-plc: abstract datatypes (CGP-310)

### PR DESCRIPTION
**Pull request by michaelpj**
This is a fairly significant change to codegen. At the moment, we
generate all types inline. This has the advantage of simplicity, and
makes constructing and destructing types nice and easy, but the
resulting code is very hard to read, and inlining everything makes it
very big as well.

Instead, we now use the "abstract data type" style of definition. This
is quite fiddly for reasons I won't repeat here - see the `Abstract data
types` note in the code for a detailed explanation of what's going on.

The resulting codegen does not look much better for small examples,
because there is still a fair amount of overhead in defining the types,
and we now need explicit constructor and matcher functions. However, I
think it will pay off a lot for bigger examples.

This also makes it easier to support hoisted definitions for non-locally
defined terms, although we don't do that yet.

### CODE REVIEW

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 1]
```diff
@@ -731,38 +937,13 @@ mkTyForall v body = do
...
 mkTyLam v body = do
     let ghcName = GHC.getName v
-    (k', t') <- convTyVarFresh v
+    (t', k') <- convTyVarFresh v
     body' <- local (\c -> c {ccScopes=pushTyName ghcName t' (ccScopes c)}) body
     pure $ PLC.TyLam () t' k' body'
 
-
--- See Note [Recursion with Z]
-delayFunction :: MonadQuote m => PLCType -> PLCExpr -> m PLCExpr
```
Comment (j-mueller) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: This isn't related to the abstract datatypes implementation, right? It might be clearer to move it to a separate commit (and the next two deletions below)

Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Yeah, they're dead. I think I did have them in a separate commit at some point, but I must have rebased it in.

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+-- | A definition of type 'val' with variable type 'var'.
+data Def var val = Def {dVis::Visibility, dVar::var, dVal::val}
+
+isVisible :: Def var val -> Bool
+isVisible (Def vis _ _) = case vis of
+    Abstract -> False
+    Visible  -> True
+
+-- | Either a simple type or a datatype with constructors and a matcher.
+data TypeRep = PlainType {trType::PLCType} | DataType {trType::PLCType, trConstrs::[TermDef], trMatch::TermDef}
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: I just folded it into `tydTy`.

**Comment by michaelpj** [comment]
I'll merge this in a bit since people seem happy.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 3]
```diff
@@ -391,19 +510,22 @@ getDataCons tc' = sortConstructors tc' <$> extractDcs tc'
 -- See Note [Scott encoding of datatypes]
 convDataCons :: forall m. Converting m => GHC.TyCon -> [GHC.DataCon] -> m PLCType
 convDataCons tc dcs = do
-    abstracted <- flip (foldr (\tv acc -> mkTyLam tv acc)) (GHC.tyConTyVars tc) $ do
-        resultType <- safeFreshTyName $ (GHC.getOccString $ GHC.getName tc) ++ "_matchOut"
-        cases <- mapM (dataConCaseType (PLC.TyVar () resultType)) dcs
-        pure $ mkScottTyBody resultType cases
+    abstracted <- flip (foldr (\tv acc -> mkTyLam tv acc)) (GHC.tyConTyVars tc) $ mkScottTy tc dcs
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Add/use `mkIterTyLam`?

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 4]
```diff
@@ -391,19 +510,22 @@ getDataCons tc' = sortConstructors tc' <$> extractDcs tc'
...
 
-    tds <- gets csTypeDefs
+    tcDefs <- gets csTypeDefs
     isSelfRecursive <- isSelfRecursiveTyCon tc
     if isSelfRecursive
-    then case Map.lookup (GHC.getName tc) tds of
-        Just (InProgress tyname) -> pure $ PLC.TyFix () tyname abstracted
-        _                        -> throwPlain $ ConversionError "Could not find var to fix over"
+    then case Map.lookup (GHC.getName tc) tcDefs of
+        Just (Def{dVis=Abstract,dVar=(tyname, _)}, _) -> pure $ PLC.TyFix () tyname abstracted
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: So before definitions are constructed, they're abstract, then they're visible and then they're abstract again? That's a bit confusing. Or am I reading this wrongly?

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 5]
```diff
@@ -424,48 +546,62 @@ dataConCaseType resultType dc = withContextM (sdToTxt $ "Converting data constru
...
-                                Just (pf, n) -> pure $ Just (pf, n)
-                                Nothing -> throwPlain $ ConversionError "Could not compute pattern functor for recursive type"
-                       else pure Nothing
-
-            let scottBody = mkScottConstructorBody resultType (zip caseArgNames caseTypes) (zip argNames argTypes) index maybePf
-            pure scottBody
+        withContextM (sdToTxt $ "Converting data constructor:" GHC.<+> GHC.ppr dc) $
+            -- no need for a body value check here, we know it's a lambda (see Note [Value restriction])
+            -- /\ tv_1 .. tv_n . body
+            flip (foldr (\tv acc -> mkTyAbs tv acc)) tcTyVars $ do
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Add/use `mkIterTyAbs`? And perhaps give it (and others?) a different name, so they're not confused with stuff from `MkPlc`?

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 6]
```diff
@@ -424,48 +546,62 @@ dataConCaseType resultType dc = withContextM (sdToTxt $ "Converting data constru
...
-                       then do
-                           tc' <- convTyCon tc
-                           case splitPatternFunctor tc' of
-                                Just (pf, n) -> pure $ Just (pf, n)
-                                Nothing -> throwPlain $ ConversionError "Could not compute pattern functor for recursive type"
-                       else pure Nothing
-
-            let scottBody = mkScottConstructorBody resultType (zip caseArgNames caseTypes) (zip argNames argTypes) index maybePf
-            pure scottBody
+        withContextM (sdToTxt $ "Converting data constructor:" GHC.<+> GHC.ppr dc) $
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Very nice idea in general.

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+wrapWithDefs
+    :: (MonadError ConvError m)
+    => DefMap GHC.Name TypeDef
+    -> DefMap GHC.Name TermDef
+    -> PLC.Term PLC.TyName PLC.Name ()
+    -> m (PLC.Term PLC.TyName PLC.Name ())
+wrapWithDefs typeDefs termDefs body = do
+    let sccs = defSccs typeDefs termDefs
+    -- process from the inside out
+    foldM wrapDefScc body (reverse sccs)
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: `reverse` something that has just been `reverse`d?

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+wrapDefScc
+    :: (MonadError ConvError m)
+    => PLC.Term PLC.TyName PLC.Name ()
+    -> Graph.SCC TermOrTypeDef
+    -> m (PLC.Term PLC.TyName PLC.Name ())
+wrapDefScc acc scc = case scc of
+    Graph.AcyclicSCC def            -> pure $ wrapDef acc def
+    -- self-recursive types are okay, but we don't handle recursive groups of definitions in general at the moment
+    Graph.CyclicSCC [def@TypeDef{}] -> pure $ wrapDef acc def
+    Graph.CyclicSCC _               -> throwPlain $ UnsupportedError "Recursive definitions not currently supported"
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: Should it be `"Mutually recursive definitions not currently supported"`?

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+-- | Wrap a term with a single definition.
+wrapDef :: PLC.Term PLC.TyName PLC.Name () -> TermOrTypeDef -> PLC.Term PLC.TyName PLC.Name ()
+wrapDef term def = case def of
+    TypeDef d ->
+        -- See Note [Abstract data types]
+        let
+            constructors = join $ toList $ tydConstrs d
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: I think, `fromMaybe []` is both shorter and more readable.

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            -- we don't bother binding things that are not abstract since they will have
+            -- been inlined
+            abstractTys = filter (not . isVisible) [d]
+            abstractTerms = filter (not . isVisible) (constructors ++ destructors)
+            tyVars = fmap dVar abstractTys
+            tys = fmap (trTy . dVal) abstractTys
+            vars = fmap dVar abstractTerms
+            vals = fmap dVal abstractTerms
+        in
+            mkIterApp (mkIterInst (mkIterTyAbs tyVars (mkIterLamAbs vars term)) tys) vals
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: Super nice.

File: core-to-plc/src/Language/Plutus/CoreToPLC/PLCTypes.hs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+module Language.Plutus.CoreToPLC.PLCTypes where
+
+import qualified Language.PlutusCore as PLC
+
+type PLCKind = PLC.Kind ()
+type PLCType = PLC.Type PLC.TyName ()
+type PLCTerm = PLC.Term PLC.TyName PLC.Name ()
+
+type PLCTyVar = (PLC.TyName (), PLCKind)
+type PLCVar = (PLC.Name (), PLCType)
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC/PLCTypes.hs]: I'd prefer a `data` in both the cases. Easier to read pattern matches.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 12]
```diff
@@ -567,7 +773,7 @@ convOrdMethod name
     | GHC.getOccString name == "<" = pure $ mkIntRel PLC.LessThanInteger
     | otherwise = throwSd UnsupportedError $ "Ord method:" GHC.<+> GHC.ppr name
 
-convNumMethod :: (Converting m) => GHC.Name -> m PLCExpr
+convNumMethod :: (Converting m) => GHC.Name -> m PLCTerm
 convNumMethod name
     -- only this one has a name defined in the lib??
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: It has just occurred to me that `-` is a rather unique thing: it's treated sometimes like a prefix operator (the only in Haskell?) and sometimes like an infix operator, so no surprise it has a name when others don't. Not sure whether this was said before.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 13]
```diff
@@ -487,7 +623,77 @@ mkScottConstructorBody resultTypeName caseNamesAndTypes argNamesAndTypes index m
...
+getMatchInstantiated = \case
+    (GHC.splitTyConApp_maybe -> Just (tc, args)) -> do
+        match <- getMatch tc
+
+        args' <- mapM convType args
+        pure $ mkIterInst match args'
+    -- must be a TC app
+    _ -> throwPlain $ ConversionError "Type was not a type constructor application"
+
+convAlt :: Converting m => Bool -> GHC.DataCon -> GHC.CoreAlt -> m PLCTerm
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: This function and several above are not commented. Really hard to review.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 14]
```diff
@@ -317,6 +363,79 @@ type of the overall case expression.
...
+<defn of c_n>
+<defn of match>
+(see Note [Scott-encoding of datatypes] for how the actual definitions are constructed)
+
+We also have to convert data constructors and case expressions into suitably instantiated
+calls to the constructors/destructors instead of converting them directly as described in
+Note [Scott encoding of data types]
+
+The requirements imposed by 3. render translation somewhat tricky. We want our translation to be
+ambivalent about where we are in the process of creating a type, so we use a richer type
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: I'd appreciate an explanation of why we need the ambivalence. I.e. which parts of the code work with both the visible and abstract data types?

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs [hunk 15]
```diff
@@ -8,8 +8,10 @@
 module Language.Plutus.CoreToPLC where
```
Comment (effectfully) [inline core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Maybe split this file somehow? It's so big.

Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Yeah, I've tried a few times. Unfortunately converting types and terms ends up being mutually recursive. I think I might just bite the bullet and add some `.hs-boot` files to make it work.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs (hunk 3 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Yeah, probably worth it.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs (hunk 4 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Yes, that's right. If you read the note carefully, you'll see that in the first stage we bind them to the variable that we will fix over. This is a bit I'm not totally happy with, but we represent that as an abstract definition of a type name.

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs (hunk 7 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: Yes, this is weird, but I think it's clearer to produce them in the right order and then reverse them where we "really" reverse them. I'm hoping I'll be able to switch to a more sensible graph library, at which point there will be only one reverse.

File: core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs (hunk 8 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC/Definitions.hs]: Yes.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs (hunk 13 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: Yes, I should document these more.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs (hunk 14 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: All of it! In particular conversion of types, where we want to handle e.g.
```
data List a = Nil | Cons a (List a) -- here `List` is the recursive variable
data ListWrapper a = ListWrapper (List a) -- here `List` is a reference to the opaque type
```
without the code that handles the types knowing that anything is different. Which works very nicely here. I'll add a comment.

**Comment by michaelpj** [comment]
Thanks for the review - I didn't realise you were looking at it otherwise I would have waited (and my follow-up has got stuck anyway). I'll address your comments in another PR.

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs (hunk 4 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: I used to have a separate kind of thing for this, but it was very noisy to work with. I'm still not very happy with the way this bit in particular works (finding which variable to fix over).

File: core-to-plc/src/Language/Plutus/CoreToPLC.hs (hunk 14 above)
Comment (effectfully) [reply core-to-plc/src/Language/Plutus/CoreToPLC.hs]: See now. Thanks.
